pkg/pf/tests/internal/cross-tests/tfwrite.go

```diff
@@ -1,173 +1,182 @@
-//
-//	key {
-//	  count = 1
-//	}
-//	key {
-//	  count = 2
-//	}
-//
-// This is why writePfBlockProvider is called with parentBody, instead of with the block body
-// already created (as with [writeBlock]).
-func writePfBlockProvider(key string, parentBody *hclwrite.Body, schemas pschema.Block, value cty.Value) {
-	switch schemas := schemas.(type) {
+}
+
+func pSchemaBlockToObject(block pschema.Block) pschema.NestedBlockObject {
+	switch block := block.(type) {
 	case pschema.ListNestedBlock:
-		for _, v := range value.AsValueSlice() {
-			b := parentBody.AppendNewBlock(key, nil).Body()
-			writePfObjectProvider(b, schemas.NestedObject, v.AsValueMap())
-		}
+		return block.NestedObject
 	case pschema.SetNestedBlock:
-		values := value.AsValueSet().Values()
-		for _, v := range values {
-			b := parentBody.AppendNewBlock(key, nil).Body()
-			writePfObjectProvider(b, schemas.NestedObject, v.AsValueMap())
-		}
+		return block.NestedObject
 	case pschema.SingleNestedBlock:
-		body := parentBody.AppendNewBlock(key, nil).Body()
-
-		if value.IsNull() {
-			return
+		return pschema.NestedBlockObject{
+			Attributes: block.Attributes,
+			Blocks:     block.Blocks,
 		}
-
-		writePfObjectProvider(body, pschema.NestedBlockObject{
-			Attributes: schemas.Attributes,
-			Blocks:     schemas.Blocks,
-		}, value.AsValueMap())
 	default:
-		contract.Failf("Unknown block type: %T", schemas)
+		contract.Failf("Unknown block type: %T", block)
 	}
+	return pschema.NestedBlockObject{}
 }
 
-func writePfBlockResource(key string, parentBody *hclwrite.Body, schemas rschema.Block, value cty.Value) {
-	switch schemas := schemas.(type) {
+func rSchemaBlockToObject(block rschema.Block) rschema.NestedBlockObject {
+	switch block := block.(type) {
 	case rschema.ListNestedBlock:
-		for _, v := range value.AsValueSlice() {
-			b := parentBody.AppendNewBlock(key, nil).Body()
-			writePfObjectResource(b, schemas.NestedObject, v.AsValueMap())
-		}
+		return block.NestedObject
 	case rschema.SetNestedBlock:
-		values := value.AsValueSet().Values()
-		for _, v := range values {
-			b := parentBody.AppendNewBlock(key, nil).Body()
-			writePfObjectResource(b, schemas.NestedObject, v.AsValueMap())
-		}
+		return block.NestedObject
 	case rschema.SingleNestedBlock:
-		body := parentBody.AppendNewBlock(key, nil).Body()
-
-		if value.IsNull() {
-			return
+		return rschema.NestedBlockObject{
+			Attributes: block.Attributes,
+			Blocks:     block.Blocks,
 		}
-
-		writePfObjectResource(body, rschema.NestedBlockObject{
-			Attributes: schemas.Attributes,
-			Blocks:     schemas.Blocks,
-		}, value.AsValueMap())
 	default:
-		contract.Failf("Unknown block type: %T", schemas)
+		contract.Failf("Unknown block type: %T", block)
 	}
+	return rschema.NestedBlockObject{}
+}
+
+type hclSchemaPFProvider struct {
+	sch pschema.Schema
+}
+
+var _ crosstestsimpl.ShimHCLSchema = hclSchemaPFProvider{}
+
+func NewHCLSchemaPFProvider(sch pschema.Schema) crosstestsimpl.ShimHCLSchema {
```

Comment:
I don't believe this wrapper is necessary, but it should definitely be private.
```suggestion
func newHCLSchemaPFProvider(sch pschema.Schema) crosstestsimpl.ShimHCLSchema {
```